sw/device/lib/msg_print.c

```diff
@@ -0,0 +1,107 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/msg_print.h"
+
+#include <stdarg.h>
+
+static void print_digit(print_char_func print_char, unsigned int d) {
+  print_char("0123456789ABCDEF"[d]);
+}
+
+static void print_num(print_char_func print_char, int w, unsigned int n,
+                      unsigned int base) {
+  if (--w > 0 || n >= base) {
+    print_num(print_char, w, n / base, base);
+  }
+  print_digit(print_char, n % base);
+}
```

Comment:
This implementation is both difficult to parse, and wasteful use of the stack. Please re-write it to use a loop, instead. Also, `w` really wants to be named `width`.

Reply:
Made it inline to reduce stack usage. Will keep the implementation as is and let SW team modify / optimize it as needed (as originally agreed upon), considering this is not going to be used by DV. Added a TODO for this.